module/zfs/dnode.c

```diff
@@ -1778,15 +1778,24 @@ dnode_try_claim(objset_t *os, uint64_t object, int slots)
 }
 
 /*
- * Checks if the dnode contains any uncommitted dirty records.
+ * Checks if the dnode itself is dirty, or is carrying any uncommitted records.
+ * Its important to check both, as some operations (eg appending to a file) can
+ * dirty both as a single logical unit, but they are not synced out atomically,
+ * so checking one and not the other can result an object appearing to be clean
+ * mid-way through a commit.
+ *
+ * Do not change this lightly! If you get it wrong, dmu_offset_next() can
+ * detect a hole where there is really data, leading to silent corruption.
  */
+
 boolean_t
 dnode_is_dirty(dnode_t *dn)
 {
 	mutex_enter(&dn->dn_mtx);
 
 	for (int i = 0; i < TXG_SIZE; i++) {
-		if (multilist_link_active(&dn->dn_dirty_link[i])) {
+		if (multilist_link_active(&dn->dn_dirty_link[i]) ||
+		    list_head(&dn->dn_dirty_records[i])) {
```

Comment:
given this is locked under `dn->dn_mutex` I recon the order of these checks does not matter, right ?

Reply:
Correct. Thanks for checking :)